### PR TITLE
Improve core metadata inheritance support

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Add inheritance support to EntityType.Navigations and EntityType.GetForeignKeys()
Use the root type for key values
Expose EntityType on EntityKey
Use inherited FKs and referencing FKs when finding related entities
Make the principal entity type required on FK constructor
Rename Navigation.EntityType, Property.EntityType, Key.EntityType, Index.EntityType and ForeignKey.EntityType to DeclaringEntityType
Move IEntityType.GetProperty(string) and IEntityType.GetNavigation(string) to extension methods
Add FindForeignKey, FindIndex and FindKey to IEntityType

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
I will add more tests for the changed code when I'll add Code First inheritance support.

File: src/EntityFramework.Core/Metadata/EntityType.cs [hunk 1]
```diff
@@ -993,6 +961,16 @@ public int Compare(IReadOnlyList<Property> x, IReadOnlyList<Property> y)
...
             }
+
+            public bool Equals(IReadOnlyList<Property> x, IReadOnlyList<Property> y)
+            {
+                return Compare(x, y) == 0;
+            }
+
+            public int GetHashCode(IReadOnlyList<Property> obj)
+            {
+                return obj.Aggregate(0, (hash, p) => hash ^ p.GetHashCode());
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: Expression body

File: src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs [hunk 2]
```diff
@@ -20,6 +20,11 @@ public static IEnumerable<IEntityType> GetDerivedTypes([NotNull] this IEntityTyp
             return GetDerivedTypes(entityType.Model, entityType);
         }
 
+        public static IEnumerable<EntityType> GetDerivedTypes([NotNull] this EntityType entityType)
+        {
+            return ((IEntityType)entityType).GetDerivedTypes().Cast<EntityType>();
```
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs]: Check null

Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs]: Never mind. I see it is checked in the overload. In that case, expression body. :-) And other places here as well.

**Comment by ajcvickers** [comment]
:shipit:

**Pull request merged**